x-pack/plugins/osquery/public/shared_components/osquery_response_action_type/index.tsx

```diff
@@ -0,0 +1,222 @@
+  const { errors, isValid } = formState;
+  const context = useFormContext();
+  const data = context.getFormData();
+  const { params: defaultParams } = get(data, item.path);
+
+  const watchedValues = watch();
+  const { data: packData } = usePack({
+    packId: watchedValues?.packId?.[0],
+    skip: !watchedValues?.packId?.[0],
+  });
+  const [queryType, setQueryType] = useState<string>(
+    !isEmpty(defaultParams?.queries) ? 'pack' : 'query'
+  );
+  const onSubmit = useCallback(async () => {
+    try {
+      if (queryType === 'pack') {
+        context.updateFieldValues({
+          [item.path]: {
+            actionTypeId: OSQUERY_TYPE,
+            params: {
+              id: watchedValues.id,
+              packId: watchedValues?.packId?.length ? watchedValues?.packId[0] : undefined,
+              queries: packData
+                ? map(packData.queries, (query, queryId: string) => ({
+                    ...query,
+                    id: queryId,
+                  }))
+                : watchedValues.queries,
+            },
+          },
+        });
+      } else {
+        context.updateFieldValues({
+          [item.path]: {
+            actionTypeId: OSQUERY_TYPE,
+            params: {
+              id: watchedValues.id,
+              savedQueryId: watchedValues.savedQueryId,
+              query: watchedValues.query,
+              ecs_mapping: convertECSMappingToObject(watchedValues.ecs_mapping),
+            },
+          },
+        });
+      }
+      // eslint-disable-next-line no-empty
+    } catch (e) {}
+  }, [
+    context,
+    item.path,
+    packData,
+    queryType,
+    watchedValues.ecs_mapping,
+    watchedValues.id,
+    watchedValues?.packId,
+    watchedValues.queries,
+    watchedValues.query,
+    watchedValues.savedQueryId,
+  ]);
+
+  useEffect(() => {
+    // @ts-expect-error update types
+    if (ref && ref.current) {
+      // @ts-expect-error update types
+      ref.current.validation[item.id] = async () => {
+        await handleSubmit(onSubmit)();
+
+        return isEmpty(errors);
+      };
+    }
+  }, [errors, handleSubmit, isValid, item.id, onSubmit, ref, watchedValues]);
+
+  useEffect(() => {
+    register('savedQueryId');
+    register('id');
+  }, [register]);
+
+  const permissions = useKibana().services.application.capabilities.osquery;
+
+  useEffect(() => {
+    if (defaultParams && defaultParams.id) {
+      const { packId, ecs_mapping: ecsMapping, ...restParams } = defaultParams;
+      map(restParams, (value, key: keyof OsqueryResponseActionsParamsFormFields) => {
+        if (!isEmpty(value)) {
+          setValue(key, value);
+        }
+      });
+      if (ecsMapping) {
+        const converted = convertECSMappingToFormValue(ecsMapping);
+        setValue('ecs_mapping', converted);
+      }
+
+      if (!isEmpty(packId)) {
+        setValue('packId', [packId]);
+      }
+    }
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
```

Comment:
I get why you disable this to only run once, but this line is terrifying and often causes bugs, there are a few other ways to accomplish the run once by refactoring the hook to be something like 
```
  const mungeData = useCallback(() => {
    if (defaultParams && defaultParams.id) {
      const { packId, ecs_mapping: ecsMapping, ...restParams } = defaultParams;
      map(restParams, (value, key: keyof OsqueryResponseActionsParamsFormFields) => {
        if (!isEmpty(value)) {
          setValue(key, value);
        }
      });
      if (ecsMapping) {
        const converted = convertECSMappingToFormValue(ecsMapping);
        setValue('ecs_mapping', converted);
      }

      if (!isEmpty(packId)) {
        setValue('packId', [packId]);
      }
    }
  }, [defaultParams, setValue]);

  useEffect(() => {
    mungeData();
  }, [mungeData]);
```
the callback only runs when the reference in the useEffect changes, which won't happen just from render cycles

Reply:
Changed it into useEffectOnce from react-use/lib/useEffectOnce , hope this is better :)